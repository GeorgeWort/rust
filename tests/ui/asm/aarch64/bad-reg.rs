// only-aarch64
// compile-flags: -A incomplete-features -C target-feature=+neon,+sve

#![feature(repr_simd, stdarch_aarch64_sve, stdsimd, asm_const, unsized_locals, unsized_fn_params)]

use std::arch::aarch64::float64x2_t;
use std::arch::aarch64::sve::svfloat64_t;
use std::arch::aarch64::sve::svdup_n_f64;
use std::arch::asm;

fn main() {
    let mut foo = 0;
    let mut bar = 0;

    let f64x2: float64x2_t = unsafe { std::mem::transmute(0i128) };
    let mut f64xN: svfloat64_t = svdup_n_f64(0.0);

    unsafe {
        // Bad register/register class

        asm!("{}", in(foo) foo);
        //~^ ERROR invalid register class `foo`: unknown register class
        asm!("", in("foo") foo);
        //~^ ERROR invalid register `foo`: unknown register
        asm!("{:z}", in(reg) foo);
        //~^ ERROR invalid asm template modifier for this register class
        asm!("{:r}", in(vreg) foo);
        //~^ ERROR invalid asm template modifier for this register class
        asm!("{:r}", in(vreg_low16) foo);
        //~^ ERROR invalid asm template modifier for this register class
        asm!("{:v}", in(zreg) foo);
        //~^ ERROR invalid asm template modifier for this register class
        asm!("{:a}", const 0);
        //~^ ERROR asm template modifiers are not allowed for `const` arguments
        asm!("{:a}", sym main);
        //~^ ERROR asm template modifiers are not allowed for `sym` arguments
        asm!("", in("x29") foo);
        //~^ ERROR invalid register `x29`: the frame pointer cannot be used as an operand
        asm!("", in("sp") foo);
        //~^ ERROR invalid register `sp`: the stack pointer cannot be used as an operand
        asm!("", in("xzr") foo);
        //~^ ERROR invalid register `xzr`: the zero register cannot be used as an operand
        asm!("", in("x19") foo);
        //~^ ERROR invalid register `x19`: x19 is used internally by LLVM and cannot be used as an operand for inline asm

        asm!("{}", in(zreg) f64x2);
        //~^ ERROR type `float64x2_t` cannot be used with this register class
        asm!("{}", in(vreg) f64xN);
        //~^ ERROR type `svfloat64_t` cannot be used with this register class

        asm!("", in("ffr") foo);
        //~^ ERROR register class `ffr_reg` can only be used as a clobber, not as an input or output
        //~| ERROR type `i32` cannot be used with this register class
        asm!("", out("ffr") _);
        asm!("{}", in(ffr_reg) foo);
        //~^ ERROR register class `ffr_reg` can only be used as a clobber, not as an input or output
        //~| ERROR type `i32` cannot be used with this register class
        asm!("{}", out(ffr_reg) _);
        //~^ ERROR register class `ffr_reg` can only be used as a clobber, not as an input or output

        // Explicit register conflicts
        // (except in/lateout which don't conflict)

        asm!("", in("x0") foo, in("w0") bar);
        //~^ ERROR register `x0` conflicts with register `x0`
        asm!("", in("x0") foo, out("x0") bar);
        //~^ ERROR register `x0` conflicts with register `x0`
        asm!("", in("w0") foo, lateout("w0") bar);
        asm!("", in("v0") foo, in("q0") bar);
        //~^ ERROR register `v0` conflicts with register `v0`
        asm!("", in("v0") foo, out("q0") bar);
        //~^ ERROR register `v0` conflicts with register `v0`
        asm!("", in("v0") foo, lateout("q0") bar);
        asm!("", in("v0") foo, in("z0") bar);
        //~^ ERROR register `z0` conflicts with register `v0`
        asm!("", in("v0") foo, out("z0") bar);
        //~^ ERROR register `z0` conflicts with register `v0`
        asm!("", in("v0") foo, lateout("z0") bar);
    }
}
