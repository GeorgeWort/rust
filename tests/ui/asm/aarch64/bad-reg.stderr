error: invalid register class `foo`: unknown register class
  --> $DIR/bad-reg.rs:21:20
   |
LL |         asm!("{}", in(foo) foo);
   |                    ^^^^^^^^^^^

error: invalid register `foo`: unknown register
  --> $DIR/bad-reg.rs:23:18
   |
LL |         asm!("", in("foo") foo);
   |                  ^^^^^^^^^^^^^

error: invalid asm template modifier for this register class
  --> $DIR/bad-reg.rs:25:15
   |
LL |         asm!("{:z}", in(reg) foo);
   |               ^^^^   ----------- argument
   |               |
   |               template modifier
   |
   = note: the `reg` register class supports the following template modifiers: `w`, `x`

error: invalid asm template modifier for this register class
  --> $DIR/bad-reg.rs:27:15
   |
LL |         asm!("{:r}", in(vreg) foo);
   |               ^^^^   ------------ argument
   |               |
   |               template modifier
   |
   = note: the `vreg` register class supports the following template modifiers: `b`, `h`, `s`, `d`, `q`, `v`

error: invalid asm template modifier for this register class
  --> $DIR/bad-reg.rs:29:15
   |
LL |         asm!("{:r}", in(vreg_low16) foo);
   |               ^^^^   ------------------ argument
   |               |
   |               template modifier
   |
   = note: the `vreg_low16` register class supports the following template modifiers: `b`, `h`, `s`, `d`, `q`, `v`

error: invalid asm template modifier for this register class
  --> $DIR/bad-reg.rs:31:15
   |
LL |         asm!("{:v}", in(zreg) foo);
   |               ^^^^   ------------ argument
   |               |
   |               template modifier
   |
   = note: the `zreg` register class supports the following template modifiers: `b`, `h`, `s`, `d`, `q`, `z`

error: asm template modifiers are not allowed for `const` arguments
  --> $DIR/bad-reg.rs:33:15
   |
LL |         asm!("{:a}", const 0);
   |               ^^^^   ------- argument
   |               |
   |               template modifier

error: asm template modifiers are not allowed for `sym` arguments
  --> $DIR/bad-reg.rs:35:15
   |
LL |         asm!("{:a}", sym main);
   |               ^^^^   -------- argument
   |               |
   |               template modifier

error: invalid register `x29`: the frame pointer cannot be used as an operand for inline asm
  --> $DIR/bad-reg.rs:37:18
   |
LL |         asm!("", in("x29") foo);
   |                  ^^^^^^^^^^^^^

error: invalid register `sp`: the stack pointer cannot be used as an operand for inline asm
  --> $DIR/bad-reg.rs:39:18
   |
LL |         asm!("", in("sp") foo);
   |                  ^^^^^^^^^^^^

error: invalid register `xzr`: the zero register cannot be used as an operand for inline asm
  --> $DIR/bad-reg.rs:41:18
   |
LL |         asm!("", in("xzr") foo);
   |                  ^^^^^^^^^^^^^

error: invalid register `x19`: x19 is used internally by LLVM and cannot be used as an operand for inline asm
  --> $DIR/bad-reg.rs:43:18
   |
LL |         asm!("", in("x19") foo);
   |                  ^^^^^^^^^^^^^

error: register class `ffr_reg` can only be used as a clobber, not as an input or output
  --> $DIR/bad-reg.rs:51:18
   |
LL |         asm!("", in("ffr") foo);
   |                  ^^^^^^^^^^^^^

error: register class `ffr_reg` can only be used as a clobber, not as an input or output
  --> $DIR/bad-reg.rs:55:20
   |
LL |         asm!("{}", in(ffr_reg) foo);
   |                    ^^^^^^^^^^^^^^^

error: register class `ffr_reg` can only be used as a clobber, not as an input or output
  --> $DIR/bad-reg.rs:58:20
   |
LL |         asm!("{}", out(ffr_reg) _);
   |                    ^^^^^^^^^^^^^^

error: register `x0` conflicts with register `x0`
  --> $DIR/bad-reg.rs:64:32
   |
LL |         asm!("", in("x0") foo, in("w0") bar);
   |                  ------------  ^^^^^^^^^^^^ register `x0`
   |                  |
   |                  register `x0`

error: register `x0` conflicts with register `x0`
  --> $DIR/bad-reg.rs:66:32
   |
LL |         asm!("", in("x0") foo, out("x0") bar);
   |                  ------------  ^^^^^^^^^^^^^ register `x0`
   |                  |
   |                  register `x0`
   |
help: use `lateout` instead of `out` to avoid conflict
  --> $DIR/bad-reg.rs:66:18
   |
LL |         asm!("", in("x0") foo, out("x0") bar);
   |                  ^^^^^^^^^^^^

error: register `v0` conflicts with register `v0`
  --> $DIR/bad-reg.rs:69:32
   |
LL |         asm!("", in("v0") foo, in("q0") bar);
   |                  ------------  ^^^^^^^^^^^^ register `v0`
   |                  |
   |                  register `v0`

error: register `v0` conflicts with register `v0`
  --> $DIR/bad-reg.rs:71:32
   |
LL |         asm!("", in("v0") foo, out("q0") bar);
   |                  ------------  ^^^^^^^^^^^^^ register `v0`
   |                  |
   |                  register `v0`
   |
help: use `lateout` instead of `out` to avoid conflict
  --> $DIR/bad-reg.rs:71:18
   |
LL |         asm!("", in("v0") foo, out("q0") bar);
   |                  ^^^^^^^^^^^^

error: register `z0` conflicts with register `v0`
  --> $DIR/bad-reg.rs:74:32
   |
LL |         asm!("", in("v0") foo, in("z0") bar);
   |                  ------------  ^^^^^^^^^^^^ register `z0`
   |                  |
   |                  register `v0`

error: register `z0` conflicts with register `v0`
  --> $DIR/bad-reg.rs:76:32
   |
LL |         asm!("", in("v0") foo, out("z0") bar);
   |                  ------------  ^^^^^^^^^^^^^ register `z0`
   |                  |
   |                  register `v0`
   |
help: use `lateout` instead of `out` to avoid conflict
  --> $DIR/bad-reg.rs:76:18
   |
LL |         asm!("", in("v0") foo, out("z0") bar);
   |                  ^^^^^^^^^^^^

error: type `float64x2_t` cannot be used with this register class
  --> $DIR/bad-reg.rs:46:29
   |
LL |         asm!("{}", in(zreg) f64x2);
   |                             ^^^^^
   |
   = note: register class `zreg` supports these types: i16, i32, i64, f32, f64, i8x16xN, i16x8xN, i32x4xN, i64x2xN, f32x4xN, f64x2xN
   = help: consider using the `vreg` register class instead

error: type `svfloat64_t` cannot be used with this register class
  --> $DIR/bad-reg.rs:48:29
   |
LL |         asm!("{}", in(vreg) f64xN);
   |                             ^^^^^
   |
   = note: register class `vreg` supports these types: i8, i16, i32, i64, f32, f64, i8x8, i16x4, i32x2, i64x1, f32x2, f64x1, i8x16, i16x8, i32x4, i64x2, f32x4, f64x2
   = help: consider using the `zreg` register class instead

error: type `i32` cannot be used with this register class
  --> $DIR/bad-reg.rs:51:28
   |
LL |         asm!("", in("ffr") foo);
   |                            ^^^
   |
   = note: register class `ffr_reg` supports these types: 

error: type `i32` cannot be used with this register class
  --> $DIR/bad-reg.rs:55:32
   |
LL |         asm!("{}", in(ffr_reg) foo);
   |                                ^^^
   |
   = note: register class `ffr_reg` supports these types: 

error: aborting due to 25 previous errors

